src/service/resources-impl.js

```diff
@@ -407,6 +410,12 @@ export class Resources {
    * @private
    */
   buildReadyResources_() {
+    // Avoid cases where elements add more elements inside of them
+    // and cause an infinite loop of building - see #3354 for details.
+    if (this.isCurrentlyBuildingPendingResources_) {
+      return;
+    }
+    this.isCurrentlyBuildingPendingResources_ = true;
```

Comment:
Also test these states in some of the existing tests and maybe throw one in that throws an exception.